src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+>
+> 1. **Assert**: Type(O) is Object.
+> 2. **Assert**: IsPropertyKey(P) is true.
+> 3. Let desc be **?** O.**\[\[GetOwnProperty\]\]**(P).
+> 4. If desc is undefined, return false.
+> 5. Return true.
+
+[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)
+
+But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?
+
+Let's find out!
+
+## ECMAScript Language Types and ECMAScript Specification Types
+
+Let's start with something that looks familiar. The spec uses values such as undefined, true and false, which we already know from JavaScript. They all are **ECMAScript language values**, values of **ECMAScript language types** which the spec also defines.
+
+The spec also uses ECMAScript language values internally, for example, an internal data type might contain a field whose possible values are true and false.
+
+[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)
+
+**ECMAScript specification types** are types that occur only in the ECMAScript spec, but not in the JavaScript language, and which might or might not occur as internal types in a JavaScript engine. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).
+
+[Spec: ECMAScript specification types](https://tc39.es/ecma262/#sec-ecmascript-specification-types)
+
+## Abstract operations
+
+**Abstract operations** are functions defined in the ECMAScript spec; they are defined for the purpose of writing the spec elegantly and concisely. A JavaScript engine doesn't have to implement them as separate functions inside the engine. They cannot be directly called from JavaScript.
+
+[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)
+
+## Internal slots and internal methods
+
+**Internal slots** and **internal methods** use names enclosed in \[\[ \]\].
+
+Internal slots are (assumed) data members (of a JavaScript object or a spec-internal data type) and internal methods are (assumed) member functions (of a JavaScript object or a spec-internal data type).
+
+Internal slots and methods are used in the algorithms described by the spec. Internal slots are used for storing the state of the object, and internal methods are functions associated with the object.
+
+[Spec: Object internal methods and internal slots](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots)
+
+For example, every JavaScript object has an internal slot \[\[Prototype\]\] and an internal method \[\[GetOwnProperty\]\].
+
+Internal slots and methods are not accessible from JavaScript, for example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.
+
+Each internal method delegates to a similarly-named abstract operation:
```

Comment:
This is sometimes true for ordinary objects, but many internal methods are specified with algorithmic steps like abstract operations. Perhaps "Sometimes internal methods delegate to similarly-named abstract operations, such as in the case of ordinary objects' [[GetOwnProperty]]:"?